### PR TITLE
Fix for umount02

### PR DESCRIPTION
**Pull request by ashokchandran**
Issue Description: This test was failing with panic due to lack of space. This was a limitation in SGX with 32MB memory.
Solution: The test now uses root file system

### CODE REVIEW

**Review by hukoyu — APPROVED** [review]
I already reviewed this PR and @ashokchandran addressed my feedback in private repo. This PR is resubmitted to public repo. It looks good to me. @letmaik can you also review and merge if looks good. Thanks.

File: tests/ltp/buildenv.sh [hunk 1]
```diff
@@ -45,6 +46,10 @@ if [[ "$mode" == "build" ]]; then
         git apply $FORK_DISABLE_PATCH
     fi
 
+    if [ -f $UMOUNT02_PATCH ];then
```
Comment (letmaik) [inline tests/ltp/buildenv.sh]: Why would the patch file be missing?

File: tests/ltp/umount02.patch [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
++const char* opts = "mode=0777";  
+ 
+ static char long_path[PATH_MAX + 2];
+ static int mount_flag;
+@@ -31,7 +36,7 @@ static struct tcase {
+ 	int exp_errno;
+ } tcases[] = {
+ 	{"Already mounted/busy", MNTPOINT, EBUSY},
+-	{"Invalid address", NULL, EFAULT},
++/*	{"Invalid address", NULL, EFAULT}, This test is commented because in SGX accessing NULL results in enclave exit */
```
Comment (letmaik) [inline tests/ltp/umount02.patch]: Is that a bug in SGX-LKL? Should it be recorded in an issue?

File: tests/ltp/umount02.patch [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ 
+ 	if (TST_RET != -1) {
+-		tst_res(TFAIL, "umount() succeeds unexpectedly");
++		tst_res(TFAIL, "umount() succeeds unexpectedly. TEST_FAILED");
+ 		return;
+ 	}
+ 
+ 	if (tc->exp_errno != TST_ERR) {
+-		tst_res(TFAIL | TTERRNO, "umount() should fail with %s",
++		tst_res(TFAIL | TTERRNO, "umount() should fail with %s. TEST_FAILED",
```
Comment (letmaik) [inline tests/ltp/umount02.patch]: Why is this needed? Is the exit code of the test not enough?

File: tests/ltp/umount02.patch [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
++
++	const char* src  = "/dev/vda";
++	const char* type = "ext4";
++	const unsigned long mntflags = 0;
++	const char* opts = "mode=0777";  
+ 	memset(long_path, 'a', PATH_MAX + 1);
+ 
+ 	SAFE_MKDIR(MNTPOINT, 0775);
+-	SAFE_MOUNT(tst_device->dev, MNTPOINT, tst_device->fs_type, 0, NULL);
++/*	SAFE_MOUNT(tst_device->dev, MNTPOINT, tst_device->fs_type, 0, NULL);*/
```
Comment (letmaik) [inline tests/ltp/umount02.patch]: Use `//` when only commenting out a single line. And maybe also add a comment why the patch to mount is necessary.

File: tests/ltp/umount02.patch [hunk 5]
```diff
@@ -0,0 +1,81 @@
+diff --git a/testcases/kernel/syscalls/umount/umount02.c b/testcases/kernel/syscalls/umount/umount02.c
+index cd67aa6d2..8822eab89 100644
+--- a/testcases/kernel/syscalls/umount/umount02.c
++++ b/testcases/kernel/syscalls/umount/umount02.c
+@@ -20,6 +20,11 @@
+ #include "tst_test.h"
+ 
+ #define MNTPOINT	"mntpoint"
++const char* src  = "/dev/vda";
```
Comment (letmaik) [inline tests/ltp/umount02.patch]: Add a comment somewhere that explains why this patch is necessary so that it can be understood in the future.

**Review by letmaik — CHANGES_REQUESTED** [review]

File: tests/ltp/buildenv.sh (hunk 1 above)
Comment (ashokchandran) [reply tests/ltp/buildenv.sh]: Chance is less but if the patch file is not copied to the proper path we can find it here during the build.

Comment (letmaik) [reply tests/ltp/buildenv.sh]: Sorry I don't understand, could you explain it a bit more?

File: tests/ltp/umount02.patch (hunk 3 above)
Comment (ashokchandran) [reply tests/ltp/umount02.patch]: This was for a pipe line script to find result as per Huseyn.

File: tests/ltp/umount02.patch (hunk 2 above)
Comment (ashokchandran) [reply tests/ltp/umount02.patch]: Yes  will raise this as an issue..  in umount02 when it test umount with null, it causes enclave to abort. Normally application is expected to get EFAULT. But the back trace is showing that in lkl it calls oe_abort when address is invalid.
 
#0  lkl_access_ok (addr=0, size=1) at lkl/setup.c:1499
#1  0x00007fff0010aaff in __lkl_access_ok (size=<optimized out>, addr=<optimized out>) at ./arch/lkl/include/asm/uaccess.h:14
#2  strncpy_from_user (count=<optimized out>, src=<optimized out>, dst=<optimized out>) at ./include/asm-generic/uaccess.h:233
#3  getname_flags (filename=0x0, flags=0, empty=0x0) at fs/namei.c:149
#4  0x00007fff0010ad2d in getname (filename=<optimized out>) at fs/namei.c:209
#5  0x00007fff0010da1d in user_path_mountpoint_at (dfd=0, name=<optimized out>, flags=1049890816, path=0x7fff3e941020)
    at fs/namei.c:2745
#6  0x00007fff0012003d in ksys_umount (name=0x7fff3e941000 " \020\224>\377\177", flags=0) at fs/namespace.c:1680
#7  0x00007fff00120479 in __do_sys_umount (flags=<optimized out>, name=<optimized out>) at fs/namespace.c:1706
#8  __se_sys_umount (name=<optimized out>, flags=<optimized out>) at fs/namespace.c:1704
#9  0x00007fff00088076 in run_syscall (params=<optimized out>, no=<optimized out>) at arch/lkl/kernel/syscalls.c:44
#10 lkl_syscall (no=140734243278880, params=0x0) at arch/lkl/kernel/syscalls.c:183
#11 0x00007fff0054901a in __filter_syscall2 (n=39, a1=0, a2=0) at ./src/internal/syscall.h:161
#12 0x00007fff0054924a in umount (special=0x0) at src/linux/mount.c:11
    if (!ret)
    {
        sgxlkl_fail("lkl_access_check failed: %p\n", (void*)addr);
    }
    return ret;

File: tests/ltp/umount02.patch (hunk 3 above)
Comment (letmaik) [reply tests/ltp/umount02.patch]: @hukoyu Can you comment?

File: tests/ltp/buildenv.sh (hunk 1 above)
Comment (ashokchandran) [reply tests/ltp/buildenv.sh]: I meant copying patch to mount point is done in Makefile, So thought its better to check here that is there. 
vamsi@vm-gshanker:~/ashok_latest/fork3/sgx-lkl/tests/ltp$ grep patch Makefile
        $(ESCALATE_CMD) cp ltp_fork_disable.patch $(MOUNTPOINT)/
        $(ESCALATE_CMD) cp patches/umount03.patch $(MOUNTPOINT)/

File: tests/ltp/umount02.patch (hunk 4 above)
Comment (ashokchandran) [reply tests/ltp/umount02.patch]: will do

File: tests/ltp/umount02.patch (hunk 5 above)
Comment (ashokchandran) [reply tests/ltp/umount02.patch]: will do

File: tests/ltp/buildenv.sh (hunk 1 above)
Comment (letmaik) [reply tests/ltp/buildenv.sh]: So if the patch is not there then you don't apply it and continue happily as if everything is ok? Doesn't make much sense to me. Or am I missing something?

Comment (ashokchandran) [reply tests/ltp/buildenv.sh]: Agree that it is not really required. Will remove the check.

**Comment by ashokchandran** [comment]
@letmaik  please review and approve if all ok now.

**Review by letmaik — APPROVED** [review]
